### PR TITLE
Add pub and stage membership management

### LINKED ISSUE #594: Add basic Pub membership UI
## Motivation

- So admins can assign Members to individual pubs
- So admins can create reviews of pubs as children or siblings that their submitters, as contributors, can't see (relies on conceit that any form/editor that creates children will add the creator as admins of those child pubs -- seems reasonable?)
- So admins can assign editors who are responsible for managing pubs and their children without giving them deletion permissions

## Requirements

- Ability to add/remove existing community members to a Pub as an admin, contributor, or editor in the stage member tab, and invite new ones.
  - Add "add member" button to member section of the sidebar, identical to "members" page add member button
  - Only admins should see the button
  - Should repurpose the existing member add UI, e.g.:
    - When clicked, shown a popup to search by email
    - If user found, can select permission for the stage and add it
    - If no user found, sends an invitation. That invitation should use slightly different language, e.g.: `You have been invited to be a [role name] of the [stage name] stage in [community name].`
- Ability to see existing members and their role, and delete them in the member section of the pub sidebar
  - Only admins should be able to see the delete button
  - Can just be the current member list, plus their role underneath their name and an x to delete them, similar to the current pub type editor

## Acceptance Criteria
- Does not implement permissions, that's a separate card, but should add/remove the correct permissions
- Does not need to show cascading permissions yet, we can add that later

## Mockups (if available)
Image

<details>
  <summary>Old</summary>

Permissions behavior:
- Contributor:
  - Can see and edit pub fields
  - Can see what stage a pub is in
  - Can create child/related Pubs (they become direct members of these pubs)
  - Cannot take actions, move, or assign
- Editor: same as contributor, plus:
  - Can see and edit children and siblings (even if not directly added)
  - Can move, assign, and run actions on pubs (even if they're not stage members)
- Admin: same as editor, plus:
  - Can delete the pub
  - Can add members to the pub as contributor, editor, or community admin. (If the user does not exist, admin can invite them via email. The user receives an invitation to join PubPub as a result, with a message specifying the stage they have been invited to and a redirect to the stage configuration they have been added to.)

</details>

### LINKED ISSUE #595: Add basic Stage membership UI
## Motivation

So admins can add, remove, and change permissions of Members in stages

## Requirements

- Ability to add/remove existing community members to a stage as an admin, contributor, or editor in the stage member tab, and invite new ones.
  - Add "add member" button to member tab of stage management, identical to "members" page add member button
  - Only admins should see the button
  - Should repurpose the existing member add UI, e.g.:
    - When clicked, shown a popup to search by email
    - If user found, can select permission for the stage and add it
    - If no user found, sends an invitation. That invitation should use slightly different language, e.g.: `You have been invited to be a [role name] of the [stage name] stage in [community name].`
- Ability to see existing members and their role, and delete them in the member tab of the side panel
  - Only admins should be able to see the delete button
  - Can just be the current member list, plus their role underneath their name and an x to delete them, similar to the current pub type editor

## Acceptance Criteria
- If user is not already a member of the community, only adds a membership to the stage with the selected role — not an additional community-level role
- Does not implement permissions/querying, that's separate cards, but should add/remove the correct permissions
- Does not need to show cascading permissions yet, we can add that later

## Mockups (if available)
Image

<details>
  <summary>Old Requirements rolled into other tickets</summary>
   
- Permissions behavior:
   - Contributor: 
      - Can see and edit Pubs directly in the stage (e.g. not child/related pubs not in the stage they don't have direct access to)
      - Can create new Pubs in the stage (they become direct members of these pubs)
      - Can create stageless child/related Pubs of Pubs in the stage (they become direct members of these pubs)
      - Can run actions/move/assign on Pubs in the stage
   - Editor: same as contributor, plus: 
      - Can see and edit children/related pubs of Pubs in the stage, even if child/related Pubs are not in the stage
   - Admin: same as editor, plus: 
      - Can edit stage configuration/actions/etc.
      - Can delete pubs in the stage
      - Can add members to stage as contributor, editor, or admin. (If the user does not exist already, they can be invited via email. The user receives an invitation to join PubPub as a result, with a message specifying the stage they have been invited to and a redirect to the stage configuration they have been added to.)
- Email behavior: 
  - When a user is added to a stage from the UI, they receive an email notifying them that they have been added to that stage, with a link to visit it.

## Acceptance Criteria

- Add a new `role` column to the permissions table. This column stores an value for a new `PermissionRole` enum.
- Community level roles will continue to function as they do today. (Eventually we will likely remove community member roles and rely on a single community permission per member)
- The schema is not exactly right. Adjust the Permission table to capture the pubId, stageId, or communityId.
- Add a constraint that ensures a member may have only one permission per platform object (community, stage, pub).

</details>

### PR DESCRIPTION
**Pull request by kalilsn**
## Issue(s) Resolved
Resolves #594 
Resolves #595 

## High-level Explanation of PR
- Refactors the add member dialog to work with pub and stage memberships
- Refactors the add member dialog to no longer be controlled through url path and query parameters
	- 	Instead of fetching suggested users in a server component and rerendering, it now uses [React Query](https://tanstack.com/query/v5/docs/framework/react/quick-start) through the [ts-rest generated client](https://ts-rest.com/docs/react-query/v5-setup) to request suggestions via fetch + an api endpoint
	- This should make it more portable and more performant
	- It's no longer possible to link directly to the member add dialog, and it won't keep state after a refresh, which I think are unnecessary features. LMK if not and I can bring them back for the members page only.
- Adds the ability to change the role and remove members from the existing pub and stage member lists
- Modifies the user search endpoint to only return a user when the email matches exactly. This was always the intention from a privacy perspective, but also fixed a bug which could in some cases prevent inviting a new user if their email partially matched an existing one.

## Test Plan
- As jill admin:
   - Visit the members page and ensure you can still invite and remove members as usual
   - Visit the stage editor, open a stage, and select the stages panel.
   - Click add member and try adding a member that exists in the community
   - Click add member and try adding that member again. You shouldn't be able to
   - Click add member and try adding a member that doesn't exist in the community. Check inbucket to make sure the invite email looks ok. The signup process should work, although they won't be a member of a community so don't expect to be able to navigate beyond the settings page after logging in.
   - Make sure you can delete a member and change their role
   - Visit an individual pub and run the same tests!
- as some@pubpub.org:
    -  Visit the same individual pub that you added some users to above. You should see the members, but not be able to edit them
   
## Screenshots

Stage member management
image

Pub member management
image

Pub members in readonly mode
image

Side by side comparison of width change on pub page:
image

### CODE REVIEW

File: packages/contracts/src/resources/site.ts
Comment (kalilsn) [inline packages/contracts/src/resources/site.ts:196]: I'm not sure if making this optional was the right call, LMK if that breaks something. I assume #740 will fix it either way!

File: core/app/components/Memberships/MemberInviteForm.tsx
Comment (kalilsn) [inline core/app/components/Memberships/MemberInviteForm.tsx:89]: For some reason the form successfully detects if a user is already a member when you enter their email the first time, but if you then edit the email and then edit it back to the user's email, the form lets you add them. I handled that on the backend, so it's not a big deal, but I'm not sure why it's happening. At first I figured I missed a dependency which is why there are these extras (I will clean that up before merging since it didn't help).

Comment (kalilsn) [reply core/app/components/Memberships/MemberInviteForm.tsx:89]: I also didn't want to manage this error manually but I didn't find any other approach that worked. I tried validating it through the `rules` prop on the `<FormField>` component but apparently those don't run if you use a resolver. I also tried adding it as a `.refine` rule to the resolver dynamically in the useform call above (like we do in the `InputComponentConfigurationForm`) but I ended up with infinite rerenders that I couldn't figure out how to fix.

File: core/app/c/[communitySlug]/pubs/[pubId]/page.tsx
Comment (kalilsn) [inline core/app/c/[communitySlug]/pubs/[pubId]/page.tsx:151]: I made this wider to fit the member management interface. It's about the same width as the stage panel now

**Comment by kalilsn** [comment]
I'm not sure why the tests are failing here. They pass on my computer (including against the production build with `CI=true`) but in CI it seems like the login form isn't working. I would love some help here since I don't think I changed anything that would affect the login form!

File: core/.env.test
Comment (tefkah) [inline core/.env.test:3]: was this to debug e2e failure maybe? afaik we're not using `NEXT_PUBLIC` env vars anymore

Comment (tefkah) [reply core/.env.test:3]: that also seems to be fine

File: core/lib/api.ts
Comment (tefkah) [inline core/lib/api.ts:8]: I suggest we don't use a public env var, and just rely on the window url

```suggestion
	baseUrl: typeof window === "undefined" ? env.PUBPUB_URL : window.location.origin, 
```

this is slightly more jank, and i'm not sure the `env.PUBPUB_URL` will transfer properly or will even ever get used, but this does work. we could also just make it `? "/" : ...` or something instead of `env.PUBPUB_URL`

File: core/.env.test
Comment (kalilsn) [reply core/.env.test:3]: Yeah I added it because I'm using it, but your way sounds better so I'll fix it! Maybe this is what's making the e2e tests break? Also fwiw this change was an error, since .env.test isn't actually used when we run the playwright tests (only the vitest tests).

Comment (kalilsn) [reply core/.env.test:3]: Actually this causes the build to fail. Did you have a way around that?

File: core/app/components/Memberships/MembersList.tsx
Comment (tefkah) [inline core/app/components/Memberships/MembersList.tsx:19]: these member lists look really good! really nice you can just edit the membership inline
this is super nitpicky so feel free to ignore, but i think they'd look slightly nicer if the badge was right aligned and slightly smaller! see this and the next comment

```suggestion
				<div key={user.id} className="flex items-center justify-between gap-4">
```

File: core/app/components/Memberships/MembersList.tsx
Comment (tefkah) [inline core/app/components/Memberships/MembersList.tsx:54]:
```suggestion
					<div className="flex items-center gap-2">
						{readOnly ? (
							<span className="rounded-full bg-blue-500 px-3 py-2 text-sm font-medium capitalize text-slate-50">
								{user.role}
							</span>
						) : (
							<>
								<RoleSelect
									role={user.role}
									userId={user.id}
									targetId={targetId}
									setRole={setRole}
								/>

								<RemoveMemberButton
									userId={user.id}
									targetId={targetId}
									removeMember={removeMember}
								/>
							</>
						)}
					</div>
				</div>
```

File: core/app/components/Memberships/RoleSelect.tsx
Comment (tefkah) [inline core/app/components/Memberships/RoleSelect.tsx:33]: again tiny tiny thing, but i think it'd look better if it was a small button instead

```suggestion
				<Button
					size="sm"
					className="rounded-full bg-blue-500 capitalize text-slate-50 hover:bg-blue-600"
				>
```

File: packages/contracts/src/resources/site.ts
Comment (tefkah) [reply packages/contracts/src/resources/site.ts:196]: ideally i'd do something like `z.union([authorization, cookies])` but that's not allowed unfortunately

**Review by tefkah — CHANGES_REQUESTED** [review]
looks and works great! great job, this one seemed quite finicky

I do have a few suggestion, most of which are nits and can be ignored, except for the ones about `NEXT_PUBLIC_PUBPUB_URL`. ill have a look as to why that causes builds to fail, but using the method i suggested below i can build locally it seems

File: core/lib/api.ts
Comment (tefkah) [reply core/lib/api.ts]: this is outdated, but this does seem to build for me

File: core/.env.test
Comment (tefkah) [reply core/.env.test:3]: yes, 

```
	baseUrl: typeof window === "undefined" ? env.PUBPUB_URL : window.location.origin,
```

File: core/lib/api.ts
Comment (kalilsn) [reply core/lib/api.ts]: oh right that should work!

File: core/lib/server/user.ts
Comment (kalilsn) [reply core/lib/server/user.ts]: good catch, thanks!

**Comment by kalilsn** [comment]
Great call on the right alignment and size @tefkah !

image

I think I did all your other requested changes too and that seems to have fixed the test failures, which must have been coming from my misunderstanding about whether we could use frontend env vars (we can't)!

File: .github/workflows/e2e.yml
Comment (tefkah) [inline .github/workflows/e2e.yml:124]: very useful addition

File: packages/contracts/src/resources/site.ts
Comment (tefkah) [inline packages/contracts/src/resources/site.ts:188]: very sorry, i overlooked this in my initial review, but could we make this endpoint slightly more in line with our other endpoints?

main things i'm thinking of

- instead of `searchUsers`, make it `users.search` (?).
- similarly, change the `path` to match, eg `/users/search`. 
- don't require the `communityId` in the query object. since the api routes are scoped to a community we can derive the communityid that way (see my comment in the route)

related to the above: do you think it makes sense to have separate `/users` and `/members` endpoints? I do think well add a way to get all (community) members in a community, but should that be `GET /members` or `GET /users`? We can decide that later i think, but interested in your thoughts

File: core/app/api/v0/c/[communitySlug]/site/[...ts-rest]/route.ts
Comment (tefkah) [inline core/app/api/v0/c/[communitySlug]/site/[...ts-rest]/route.ts:233]: related to my comments on the contract (again apologies for not mentioning this earlier), I think we don't need to get the `communityId` from the query, but you can just do `const community = await findCommunityBySlug()`

Ideally we'd do some other kind of permission check (probably just that the user has some kind of membership in the community?), but we can do that after this PR gets in (or if you feel really strongly about it you can use the method implemented in #805, but i don't want to block this pr for that)

**Review by tefkah — APPROVED** [review]
Other than some nitpicks on specific API things it looks great! I would like you to change these suggestions ideally, but its not blocking! Great job!

File: packages/contracts/src/resources/site.ts
Comment (kalilsn) [reply packages/contracts/src/resources/site.ts:188]: i'm happy to do whatever you think is best with the api! i named it this way because i figured it would behave a little differently than the other get endpoints, but i think users.search vs users.get captures that well enough. and i'll make the other changes!

Comment (kalilsn) [reply packages/contracts/src/resources/site.ts:188]: i'm not really sure about members vs users. i want to modify the member schema type to store a userid in the future since not everyone will have a community membership (though I would like to change that too). but other than that i don't really know what else we'll be using these endpoints for yet

File: core/app/api/v0/c/[communitySlug]/site/[...ts-rest]/route.ts
Comment (kalilsn) [reply core/app/api/v0/c/[communitySlug]/site/[...ts-rest]/route.ts:583]: i think just checking that they're logged in is good enough for now, which is what i've done. eventually we should add a new capability for this i think

**Pull request merged**
